### PR TITLE
drivers/eeprom: Implementation of eeprom m24c16 driver

### PR DESCRIPTION
**Pull request by mcaireyr**
Tested with the zephyr eeprom test software : it is fonctionnal when the
write control pin is connected directly to the GND or connected on the
microcontroller.

Signed-off-by: Marina Cairey-Remonay <marina.caireyremonay@vossloh.com>

### CODE REVIEW

File: drivers/eeprom/eeprom_m24c16.c [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+{
+	const struct eeprom_m24c16_config *config = dev->config;
+	const size_t remainder = BIT(config->addr_width) - offset;
+
+	if (len > remainder) {
+		len = remainder;
+	}
+
+	return len;
+}
```
Comment (nvincent-vossloh) [inline drivers/eeprom/eeprom_m24c16.c:51]: I don't understand the role of this function... Is it to make sure that a read does not spill over the eeprom max address ?
In that case it looks like you already check it in `eeprom_m24c16_read` at line 105.

File: drivers/eeprom/eeprom_m24c16.c [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+	return count;
+}
+
+static int eeprom_m24c16_write_page(const struct device *dev, off_t offset,
+		const void *buf, size_t len)
+{
+	const struct eeprom_m24c16_config *config = dev->config;
+	struct eeprom_m24c16_data *data = dev->data;
+	int count = eeprom_m24c16_limit_write_count(dev, offset, len);
+	uint8_t block[config->addr_width / 8 + count];
```
Comment (nvincent-vossloh) [inline drivers/eeprom/eeprom_m24c16.c:181]: I'm surprised you don't get any warnings ! 
It looks like a dynamic allocation on the stack...

**Review by nvincent-vossloh — CHANGES_REQUESTED** [review]
Minor changes and a little bit of clarification requested.

File: drivers/eeprom/eeprom_m24c16.c (hunk 1 above)
Comment (mcaireyr) [reply drivers/eeprom/eeprom_m24c16.c:51]: Yes I agree, it looks to be the same. But in the driver eeprom_at2x, it's done like this... Maybe there is a tricky case. I prefer to leave it like that.

File: drivers/eeprom/eeprom_m24c16.c (hunk 2 above)
Comment (mcaireyr) [reply drivers/eeprom/eeprom_m24c16.c:181]: I don't get any warnings.

**Review by nvincent-vossloh — APPROVED** [review]
LGTM :tada: :rocket: :fireworks:

**Pull request merged**